docs(contributing): update the coding guideline (#599)
autowarefoundation/autoware-documentation: docs/contributing/coding-guidelines/index.md

- Write comments and documentation in English.
- Functions that are too complex (low cohesion) should be appropriately split into smaller functions (e.g. less than 40 lines in one function is recommended in the [google style guideline](https://google.github.io/styleguide/cppguide.html#Write_Short_Functions)).
- Try to minimize the use of member variables or global variables that have a large scope.
- Whenever possible, break large pull requests into smaller, manageable PRs (less than 200 lines of change is recommended in some research e.g. [here](https://opensource.com/article/18/6/anatomy-perfect-pull-request)).
- When it comes to code reviews, don't spend too much time on trivial disagreements. For details see:
  - <https://en.wikipedia.org/wiki/Law_of_triviality>
  - <https://steemit.com/programming/@emrebeyler/code-reviews-and-parkinson-s-law-of-triviality>
- Please follow the guidelines for each language.
  - [C++](./languages/cpp.md)
  - [Python](./languages/python.md)
  - [Shell script](./languages/shell-scripts.md)

## Autoware Style Guide

For Autoware-specific styles, refer to the following:

- Use the `autoware_` prefix for package names.

>   - cf. [Prefix packages with autoware\_](https://github.com/orgs/autowarefoundation/discussions/4097)

Reviewer: Maybe it's better to also add the following documentation link to the reference. The documentation is where we clarify the final decision we made and the discussion explains the background of our decisions.

https://autowarefoundation.github.io/autoware-documentation/pr-599/contributing/coding-guidelines/ros-nodes/directory-structure/#package-name

Author: Thank you for your very good feedback!! I updated in https://github.com/autowarefoundation/autoware-documentation/commit/eeee09d06bdd1ce88232bd47e8098b090bdc4bfc

- Functions that are too complex (low cohesion) should be appropriately split into smaller functions.
- Try to minimize the use of member variables or global variables that have a large scope.
- Whenever possible, break large pull requests into smaller, manageable PRs.
- When it comes to code reviews, don't spend too much time on trivial disagreements. For details see:
  - <https://en.wikipedia.org/wiki/Law_of_triviality>
  - <https://steemit.com/programming/@emrebeyler/code-reviews-and-parkinson-s-law-of-triviality>

## C++ Style Guide

As a basic rule, follow [the Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html) and [the C++ Core Guidelines](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md). The following are common areas requiring special attention:

- Use fixed-width integer types like `uint32_t` instead of `int`.
  - cf. [How to C in 2016](https://matt.sh/howto-c)
- Use `const` for variables whose values will not change after initialization, and for functions or arguments that guarantee they will not modify the object's state. This improves code safety and clarity by preventing unintended side effects.

> - When not using an index, prefer `for (const auto & e : v)`.

Reviewer 2: ```suggestion
- When not using an index, prefer `for (const auto & element : values)`.
```

Author: Thank you so much!!!! :pray:

- For index access, prefer `.at(idx)` over `[idx]` whenever possible for safety.
- Avoid using auto with `Eigen::Matrix` or `Eigen::Vector` variables, as it can lead to bugs.
- Safe memory management is crucial for accident-free autonomous vehicles, so `new`, `delete`, `free`, and raw pointers are generally not acceptable.
- When using `mutex`, use `std::lock_guard` instead of `std::lock` and `std::unlock`.
- Use assert to validate that function arguments and return values meet specific conditions or expectations.
- Be cautious when using `throw`. C++ exceptions should only be used when it's acceptable for the process to terminate if uncaught. This applies to scenarios where Autoware has no other choice but to shut down.

## ROS 2 Style Guide

As a basic rule, follow [the ROS 2 developer guide](https://docs.ros.org/en/humble/The-ROS2-Project/Contributing/Developer-Guide.html) and [the ROS nodes guidelines](./ros-nodes/class-design.md). The following are commonly pointed out areas that require attention:

> - Use `RCLCPP` instead of `printf` or `std::cout`.

Reviewer 2: The `RCLCPP` is a library name.

```suggestion
- Use `RCLCPP_XXX` (eg. `RCLCPP_INFO`) macros instead of `printf` or `std::cout` for logging.
```

Please modify below too.

Author: I am very grateful for your appropriate comments!! Thank you so much :+1:

  - Reasons include the following:
    - It allows for consistent log level management. For instance, with `RCLCPP`, you can simply set `--log_level` to adjust the log level uniformly across the application.
    - You can standardize the format using `RCUTILS_CONSOLE_OUTPUT_FORMAT`.
    - With `RCLCPP`, logs are automatically recorded to `/rosout`. These logs can be saved to a rosbag, which can then be replayed to review the log data.
- Follow [the directory structure guideline](./ros-nodes/directory-structure.md).
- If `RCLCPP_INFO` generates a large amount of logs, use `RCLCPP_INFO_THROTTLE`.
- Avoid setting default values in `declare_parameter`. Instead, specify them in a config file or launch file.

## Autoware Style Guide

For Autoware-specific styles, refer to the following:

- Use the `autoware_` prefix for package names.

> - Add implementations within the `autoware` namespace.

Reviewer 2: We might add the include naming rule: `include/autoware/XXX` instead of `include/autoware_XXX`.
I'm not sure why this is required though.

Author: Thank you so much  for your suggestion!!!
I added it in https://github.com/autowarefoundation/autoware-documentation/pull/599/commits/2538ade66d1fb694f0133f73511954bcfb38dfcf

Author: > I'm not sure why this is required though.

IMO, [this](https://github.com/autowarefoundation/autoware-documentation/blob/f93e489120c1df0f7d4ccb3ebf09a4530be4a56c/docs/contributing/coding-guidelines/ros-nodes/directory-structure.md?plain=1#L211) is the reason.

Reviewer 2: Hmm... It's very weak reason.

> Rationale: To avoid verbosity.

- Keep things consistent.
- Automate where possible, using simple checks for formatting, syntax, etc.
- Write comments and documentation in English.

> - Functions that are too complex (low cohesion) should be appropriately split into smaller functions.
> - Try to minimize the use of member variables or global variables that have a large scope.
> - Whenever possible, break large pull requests into smaller, manageable PRs.

Reviewer 3: suggest to use a specific numbers or reference for better understanding

```suggestion
- Functions that are too complex (low cohesion) should be appropriately split into smaller functions (e.g. less than 40 lines in one function is recommended in the [google style guideline](https://google.github.io/styleguide/cppguide.html#Write_Short_Functions)).
- Try to minimize the use of member variables or global variables that have a large scope.
- Whenever possible, break large pull requests into smaller, manageable PRs (less than 200 is recommended in some research e.g. [here](https://opensource.com/article/18/6/anatomy-perfect-pull-request)).
```

- When it comes to code reviews, don't spend too much time on trivial disagreements. For details see:
  - <https://en.wikipedia.org/wiki/Law_of_triviality>
  - <https://steemit.com/programming/@emrebeyler/code-reviews-and-parkinson-s-law-of-triviality>

## C++ Style Guide

> As a basic rule, follow [the Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html) and [the C++ Core Guidelines](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md). The following are common areas requiring special attention:

Reviewer: Maybe we could also refer to C++ Coding guideline page in Autoware Documentation as well:
https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/languages/cpp/

Author: Thank you!! I updated in https://github.com/autowarefoundation/autoware-documentation/pull/599/commits/0d2d9c5fa6784d29b45e153ab1b0b17c930293a9

> ## C++ Style Guide

Reviewer 4: Since C++ and ROS have their own individual pages, I think it's easy to overlook the information written in the index.

Author: Thanks! 👍 
Now I'm changing it following this comment!

https://github.com/autowarefoundation/autoware-documentation/pull/599#issuecomment-2297869052

- Avoid using auto with `Eigen::Matrix` or `Eigen::Vector` variables, as it can lead to bugs.
  - cf. [Eigen, C++11 and the auto keyword](https://eigen.tuxfamily.org/dox/TopicPitfalls.html)
- Safe memory management is crucial for accident-free autonomous vehicles, so `new`, `delete`, `malloc`, `free`, and raw pointers are generally not acceptable.
  - cf. [C++ Core Guidelines, C.149: Use unique_ptr or shared_ptr to avoid forgetting to delete objects created using new](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#c149-use-unique_ptr-or-shared_ptr-to-avoid-forgetting-to-delete-objects-created-using-new), [C++ Core Guidelines, R.10: Avoid malloc() and free()](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#r10-avoid-malloc-and-free)
- When using `mutex`, use `std::lock_guard` instead of `std::lock` and `std::unlock`.
  - cf. [C++ Core Guidelines, RAII](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rr-raii)
- Use assert to validate that function arguments and return values meet specific conditions or expectations.

> - Be cautious when using `throw`. C++ exceptions should only be used when it's acceptable for the process to terminate if uncaught. This applies to scenarios where Autoware has no other choice but to shut down.

Reviewer 5: Perhaps using `it is` (not `it's`) looks better :+1:

Author: Thank you so much for your review!! :rocket: 
I fixed in https://github.com/autowarefoundation/autoware-documentation/pull/599/commits/7f812d7b39fa2a8a4b395d5d6d20046ce67f2da8
